### DISCUSSION TITLE
Extreme Postgresql disk I/O load when a user queries the federated time stream

### DISCUSSION THREAD

**Discussion by jacotec** [mastodon/mastodon#19777] [Q&A]
Hi,

I have an issue with my tiny Mastodon instance (currently only used by me). It's running in a VM on one of my Dell servers with 6GB of RAM and 6 vCPU.

It turns out that it's a real disk I/O hog.

When I just click on "Federated timeline" in my web page, the "loading" icon occurs and after a minute or so I see a "504 timeout" in the left corner of the page.

In the background, once clicked on "Federated timeline", the nightmare starts:

![image](https://user-images.githubusercontent.com/3321655/200124215-b500cc85-9bf9-40d6-9097-c8fde8152b0e.png)

The latency of the complete datastore is skyrocketing. PgHero is showing two long queries, running for more than 10 minutes (!!!) now:

`SELECT "statuses"."id", "statuses"."updated_at" FROM "statuses" INNER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."visibility" = $1 AND "accounts"."suspended_at" IS NULL AND "accounts"."silenced_at" IS NULL AND (statuses.reply = FALSE OR statuses.in_reply_to_account_id = statuses.account_id) AND (statuses.reblog_of_id IS NULL) AND "statuses"."account_id" NOT IN ($2, $3) AND "statuses"."deleted_at" IS NULL ORDER BY "statuses"."id" DESC LIMIT $4
`

I really have a couple of VMs running which uses databeases ... Two Nextcloud instances with ten thousands of files and a couple of users, a Matrix VM ... nothing is even using a fraction of disk I/O than my single user Mastodon instance.

Can anyone help me what's wrong here and what do I need to fix this?

**Comment by Gargron** [comment]
Unfortunately it is a very heavy feature. I had to manually disable the query on my servers to cope with increased traffic, for example.

**Reply by jacotec** [reply]
Not sure what you mean with " I had to manually disable the query on my servers to cope with increased traffic, for example". On my second instance this does not happen that way.

Although I ran " RAILS_ENV=production bin/tootctl statuses remove --days 30" my database size is still 6.5GB. Might this run have caused the trouble? It removed more than 250.000 old statuses (I never issued this command before)

Finally I tried to reindex the database, but that did not work either:

```
mastodon_production=# REINDEX DATABASE mastodon_production;
WARNING:  using index "pg_toast_2618_index" despite IgnoreSystemIndexes
WARNING:  using index "pg_toast_2618_index" despite IgnoreSystemIndexes
ERROR:  could not create unique index "index_custom_emojis_on_shortcode_and_domain"
DETAIL:  Key (shortcode, domain)=(raspberrypi, fosstodon.org) is duplicated.
```

**Comment by jacotec** [comment]
Even after running "RAILS_ENV=production bin/tootctl statuses remove --days 30" the "statuses" table contains 312.000 rows ... why? There are rows dated from 2019 etc ... nothing I've commented or interacted with. Why are they still there?
